### PR TITLE
JBIDE-22821 : parse generic milestone versions

### PR DESCRIPTION
**Pull request by fbricon**
Signed-off-by: Fred Bricon <fbricon@gmail.com>

### CODE REVIEW

File: foundation/plugins/org.jboss.tools.foundation.core/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersion.java [hunk 1]
```diff
@@ -24,16 +26,19 @@
 	public SimpleHierarchicalVersion(String version) {
 		internalVersion = version;
 	}
-	
+
+	private static final Pattern MILESTONES_PATTERN = Pattern.compile("^(Alpha|Beta|AM|RC|CR)(\\d+)$");
```
Comment (nickboldt) [inline foundation/plugins/org.jboss.tools.foundation.core/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersion.java]: What about Final|GA ?

File: foundation/tests/org.jboss.tools.foundation.core.test/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersionTest.java [hunk 2]
```diff
@@ -25,7 +25,12 @@ public void testGetParentVersion() {
 		assertEquals("1.2.3.CR1-v20140211-1204",  new SimpleHierarchicalVersion("1.2.3.CR1-v20140211-1204-B52").getParentVersion().toString());
 		assertEquals("1.2.3.CR1-v20140211",  new SimpleHierarchicalVersion("1.2.3.CR1-v20140211-1204").getParentVersion().toString());
 		assertEquals("1.2.3.CR1",  new SimpleHierarchicalVersion("1.2.3.CR1-v20140211").getParentVersion().toString());
-		assertEquals("1.2.3", new SimpleHierarchicalVersion("1.2.3.CR1").getParentVersion().toString());
+		assertEquals("1.2.3.Alpha",  new SimpleHierarchicalVersion("1.2.3.Alpha1").getParentVersion().toString());
+		assertEquals("1.2.3.Beta",  new SimpleHierarchicalVersion("1.2.3.Beta20").getParentVersion().toString());
+		assertEquals("1.2.3.AM",  new SimpleHierarchicalVersion("1.2.3.AM3").getParentVersion().toString());
+		assertEquals("1.2.3.RC",  new SimpleHierarchicalVersion("1.2.3.RC4").getParentVersion().toString());
```
Comment (nickboldt) [inline foundation/tests/org.jboss.tools.foundation.core.test/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersionTest.java]: We don't use RC but if you think there's a reason to include it I won't object.

File: foundation/plugins/org.jboss.tools.foundation.core/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersion.java (hunk 1 above)
Comment (fbricon) [inline foundation/plugins/org.jboss.tools.foundation.core/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersion.java]: There's no Final2 or GA3 to consider. Final/GA versions are already properly processed, and covered by matching test cases.

File: foundation/tests/org.jboss.tools.foundation.core.test/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersionTest.java (hunk 2 above)
Comment (fbricon) [inline foundation/tests/org.jboss.tools.foundation.core.test/src/org/jboss/tools/foundation/core/properties/internal/SimpleHierarchicalVersionTest.java]: Indeed there's no reason to use it for now. But the class could in theory be used to parse version for 3rd party plugins. Given it doesn't introduce any complexity, I added it.

**Comment by nickboldt** [comment]
+1, seems legit. Push it!

**Pull request merged**